pyracf/resource/resource_admin.py

```diff
@@ -589,3 +718,43 @@ def _format_profile(self, result: dict) -> None:
 
         del result["securityResult"]["resource"]["commands"][0]["messages"]
         result["securityResult"]["resource"]["commands"][0]["profiles"] = profiles
+
+    def __validate_access_level(
+        self,
+        success: Union[str, None] = None,
+        failure: Union[str, None] = None,
+        all: Union[str, None] = None,
+    ):
+        access_levels = ["alter", "control", "read", "update"]
+        value_error_text = (
+            "Please specify 'alter', 'control', 'read', or 'update' access "
+            + "for your target attempt."
+        )
+        bad_values = []
+        if success is not None:
+            if success.lower() not in access_levels:
+                bad_values.append(success)
```

Comment:
I think you can do this to consolidate things a little bit:

```python
    if (
            success is not None
            and not isinstance(success, str)
            and success.lower() not in access_levels
    ):
        bad_values.append(str(success))
```

Also it is probably a good idea to make sure the value is a string too.

Reply:
You may even be able to turn this into a for each loop as well:

```python
    for access_level in (success, failure, all):
        if (
                access_level is not None
                and not isinstance(access_level, str)
                and access_level.lower() not in access_levels
        ):
            bad_values.append(str(access_level))
```

Reply:
That was a much better and cleaner idea. I did not include checking to see if it is a string because anything that is not a string will be cast into one by the format strings or the new code as follows:

```python3
        for attempt_argument in (success, failure, all):
            if (
                attempt_argument is not None
                and str(attempt_argument).lower() not in access_levels
            ):
                bad_values.append(attempt_argument)
```

Reply:
I still think it is still probably a good idea to to check if the value is a string just to ensure that nothing strange happens if it gets to `and str(attempt_argument).lower() not in access_levels`.

Reply:
You are correct that this does only make the code marginally safer. You can correct me if I am wrong, but virtually anything in python is convertible to a string. `ThreadPoolExecutor` for example likely does not have a to string function, so I think Python just defaults to giving you a generic string telling you what kind of object the variable is and its address.

```python
>>> from concurrent.futures import ThreadPoolExecutor
>>> executor = ThreadPoolExecutor(max_workers=10)
>>> str(executor)
'<concurrent.futures.thread.ThreadPoolExecutor object at 0x000002532315DAC0>'
>>>
```

So, the only case that I can think of where there might be some strangeness, would  be if the user provided an object whose to string method returned a valid access level.

```python
>>> class MyClass():
...     def __str__(self):
...             return "alter"
...
>>> my_object = MyClass()
>>> str(my_object)
'alter'
```

And given that Python is in no way shape or form a type safe language, it can be argued that this is not a problem at all because this aligns with "duck typing", which is is not something that I don't think we have used in this project, but it is something that is fully support by Python and I am pretty sure it is probably recommended as a Pythonic way of doing things when it makes sense to do so. So from the perspective of `__validate_access_levels()` we could say in this case that given that `MyClass` looks like a string value of `alter` when we convert it to a string using `str()`, it is a valid access level. This more or less fits the "duck typing" paradigm: "If it looks like a duck and it quacks like a duck, then it is a duck."

So, in conclusion,  I agree that the benefit we get here is marginal, and I admit that this check may not be the most Pythonic, so after looking at this closer, I am in agreement that it may actually be better to just check if the value is `None` and when we convert the value to a string using `str()`, check if it ends up being a valid access level.

Reply:
Ok, that makes a lot of sense, and I think I do agree with the theory behind that. Definitely interesting to keep in mind for future work in the area. Based on this dicussion, I have changed it back to:

```python3
        for attempt_argument in (success, failure, all):
            if (
                attempt_argument is not None
                and str(attempt_argument).lower() not in valid_access_levels
            ):
                bad_access_levels.append(attempt_argument)
```

with this philosophy in mind.